mcs/class/System/System.Net/WebRequestStream.cs

```diff
@@ -0,0 +1,436 @@
+			}
+
+			await SetHeadersAsync (false, cancellationToken).ConfigureAwait (false);
+
+			Operation.ThrowIfClosedOrDisposed (cancellationToken);
+
+			if (Operation.WriteBuffer != null && !Operation.IsNtlmChallenge) {
+				await WriteRequestAsync (cancellationToken);
+				Close ();
+			}
+		}
+
+		async Task SetHeadersAsync (bool setInternalLength, CancellationToken cancellationToken)
+		{
+			Operation.ThrowIfClosedOrDisposed (cancellationToken);
+
+			if (headersSent)
+				return;
+
+			string method = Request.Method;
+			bool no_writestream = (method == "GET" || method == "CONNECT" || method == "HEAD" ||
+					      method == "TRACE");
+			bool webdav = (method == "PROPFIND" || method == "PROPPATCH" || method == "MKCOL" ||
+				      method == "COPY" || method == "MOVE" || method == "LOCK" ||
+				      method == "UNLOCK");
+
+			if (Operation.IsNtlmChallenge)
+				no_writestream = true;
+
+			if (setInternalLength && !no_writestream && HasWriteBuffer)
+				Request.InternalContentLength = WriteBufferLength;
+
+			bool has_content = !no_writestream && (!HasWriteBuffer || Request.ContentLength > -1);
+			if (!(sendChunked || has_content || no_writestream || webdav))
+				return;
+
+			headersSent = true;
+			headers = Request.GetRequestHeaders ();
+
+			WebConnection.Debug ($"{ME} SET HEADERS: {Request.ContentLength}");
+
+			try {
+				await InnerStream.WriteAsync (headers, 0, headers.Length, cancellationToken).ConfigureAwait (false);
+				var cl = Request.ContentLength;
+				if (!sendChunked && cl == 0)
+					requestWritten = true;
+			} catch (Exception e) {
+				if (e is WebException || e is OperationCanceledException)
+					throw;
+				throw new WebException ("Error writing headers", WebExceptionStatus.SendFailure, WebExceptionInternalStatus.RequestFatal, e);
+			}
+		}
+
+		internal async Task WriteRequestAsync (CancellationToken cancellationToken)
+		{
+			Operation.ThrowIfClosedOrDisposed (cancellationToken);
+
+			WebConnection.Debug ($"{ME} WRITE REQUEST: {requestWritten} {sendChunked} {allowBuffering} {HasWriteBuffer}");
+
+			if (requestWritten)
+				return;
+
+			requestWritten = true;
+			if (sendChunked || !HasWriteBuffer)
+				return;
+
+			BufferOffsetSize buffer = GetWriteBuffer ();
+			if (buffer != null && !Operation.IsNtlmChallenge && Request.ContentLength != -1 && Request.ContentLength < buffer.Size) {
+				closed = true;
+				var throwMe = new WebException ("Specified Content-Length is less than the number of bytes to write", null,
+					WebExceptionStatus.ServerProtocolViolation, null);
+				Operation.CompleteRequestWritten (this, throwMe);
+				throw throwMe;
+			}
+
+			await SetHeadersAsync (true, cancellationToken).ConfigureAwait (false);
+
+			WebConnection.Debug ($"{ME} WRITE REQUEST #1: {buffer != null}");
+
+			Operation.ThrowIfClosedOrDisposed (cancellationToken);
+			if (buffer != null && buffer.Size > 0)
+				await InnerStream.WriteAsync (buffer.Buffer, 0, buffer.Size, cancellationToken);
```

Comment:
Add .ConfigureAwait (false)

Reply:
There is already a `.ConfigureAwait (false)` in this method that will be called unconditionally before we get here, do we really need it on each `await`?